eZ/Publish/API/Repository/Iterator/BatchIterator.php

```diff
@@ -0,0 +1,94 @@
+<?php
+
+/**
+ * @copyright Copyright (C) Ibexa AS. All rights reserved.
+ * @license For full copyright and license information view LICENSE file distributed with this source code.
+ */
+declare(strict_types=1);
+
+namespace eZ\Publish\API\Repository\Iterator;
+
+use Iterator;
+
+final class BatchIterator implements Iterator
+{
+    public const DEFAULT_BATCH_SIZE = 25;
+
+    /** @var \eZ\Publish\API\Repository\Iterator\BatchIteratorAdapter */
+    private $adapter;
+
+    /** @var Iterator|null */
+    private $innerIterator;
+
+    /** @var int */
+    private $batchSize;
+
+    /** @var int */
+    private $position;
+
+    public function __construct(
+        BatchIteratorAdapter $adapter,
+        int $batchSize = self::DEFAULT_BATCH_SIZE
+    ) {
+        $this->adapter = $adapter;
+        $this->batchSize = $batchSize;
+        $this->position = 0;
+    }
+
+    public function current()
+    {
+        if (!$this->isInitialized()) {
+            $this->initialize();
+        }
+
+        return $this->innerIterator->current();
+    }
+
+    public function next(): void
+    {
+        if (!$this->isInitialized()) {
+            $this->initialize();
+        }
+
+        ++$this->position;
+        $this->innerIterator->next();
+        if (!$this->innerIterator->valid() && ($this->position % $this->batchSize) === 0) {
+            $this->innerIterator = $this->adapter->fetch($this->position, $this->batchSize);
+        }
+    }
+
+    public function key(): int
+    {
+        return $this->position;
+    }
+
+    public function valid(): bool
+    {
+        if (!$this->isInitialized()) {
+            $this->initialize();
+        }
+
+        return $this->innerIterator->valid();
+    }
+
+    public function rewind(): void
+    {
+        $this->initialize();
+    }
+
+    public function setBatchSize(int $batchSize): void
+    {
+        $this->batchSize = $batchSize;
+    }
+
+    private function initialize(): void
+    {
+        $this->innerIterator = $this->adapter->fetch(0, $this->batchSize);
+        $this->position = 0;
```

Comment:
```suggestion
        $this->position = 0;
        $this->innerIterator = $this->adapter->fetch($this->position, $this->batchSize);
```

Reply:
You could also unpack the `IteratorAggregate` here:
```php
$iterator = $this->adapter->fetch(...);

if ($iterator instanceof IteratorAggregate) {
    $iterator = $iterator->getIterator();
}

$this->innerIterator = $iterator;
```

This way change to `SearchResult` would become unnecessary and we could accept `Traversable`s.

Reply:
> You could also unpack the IteratorAggregate here:

This is responsibility of \eZ\Publish\API\Repository\Iterator\BatchIteratorAdapter implementation.

Reply:
Ok.

If we could use the suggested code variant as above (with reversed property assignment and `fetch` using `$this->position`, then
```php
$this->innerIterator = $this->adapter->fetch($this->position, $this->batchSize);
```
Could be moved to a separate method (like `doFetch`) that will do exactly as above, since all invocations of `Adapter::fetch` will be the same.

Reply:
Fixed in https://github.com/ezsystems/ezplatform-kernel/pull/211/commits/80a96f0ebb2df56009d300d9365220555e58ac76